File: src/main/java/core/scm/specialforms/Try.java
package core.scm.specialforms;

import core.environment.Environment;
import core.evaluator.Evaluator;
import core.evaluator.Reflector;
import core.exceptions.IllegalSyntaxException;
import core.exceptions.ThrowableWrapper;
import core.scm.SCMCons;
import core.scm.SCMNil;
import core.scm.SCMSymbol;

import java.util.*;

public enum Try implements ISpecialForm {
  TRY;

  private static final Reflector REFLECTOR = new Reflector();

  private static final SCMSymbol CATCH   = SCMSymbol.intern("catch");
  private static final SCMSymbol FINALLY = SCMSymbol.intern("finally");

  @Override
  public String toString() {
    return "try";
  }

  @Override
  public Object eval(List<Object> expression, Environment env, Evaluator evaluator) {
    if (expression.isEmpty()) {
      return null;
    }
    boolean hadCatch = false;
    Map<Class, Object> catches = Collections.emptyMap();
    Map<Class, SCMSymbol> catchBindings = Collections.emptyMap();
    Object fin = null;
    List<Object> expressions = new ArrayList<>();
    /* Init and check syntax */
    for (int i = 1; i < expression.size(); i++) {
      Object e = expression.get(i);
      if ((e instanceof List) && !((List) e).isEmpty()) {
        List expr = (List)e;
        Object op = expr.get(0);
        if (FINALLY.equals(op)) {
          if (i != expression.size() - 1) {
            throw new IllegalSyntaxException("try: finally clause must be last in try expression");
          }
          if (expr.size() > 1) {
            fin = SCMCons.list(Begin.BEGIN);
            ((SCMCons)fin).addAll(expr.subList(1, expr.size()));
          }
          continue;
        } else if (CATCH.equals(op)) {
          if (expr.size() < 3) {
            throw new IllegalSyntaxException("catch: bad syntax in form: " + expr);
          }
          hadCatch = true;
          if (catches.isEmpty()) {
            catches = new HashMap<>();
            catchBindings = new HashMap<>();
          }
          Class clazz = REFLECTOR.getClazz(expr.get(1).toString());
          Object catchExpr = SCMNil.NIL;
          if (expr.size() > 3) {
            catchExpr = SCMCons.list(Begin.BEGIN);
            ((SCMCons)catchExpr).addAll(expr.subList(3, expr.size()));
          }
          catches.put(clazz, catchExpr);
          Object cb = expr.get(2);
          if (!(cb instanceof SCMSymbol)) {
            throw new IllegalSyntaxException("catch: bad binding form, expected Symbol, actual: " + cb);
          }
          catchBindings.put(clazz, (SCMSymbol) cb);
          continue;
        } else {
          if (hadCatch) {
            throw new IllegalSyntaxException("try: only catch or finally clause can follow catch in try expression");
          }
        }
      }
      expressions.add(e);
    }
    /* Now Evaluate everything */
    try {
      Object result = null;
      for (Object e : expressions) {
        result = evaluator.eval(e, env);
      }
      return result;
    } catch (Throwable e) {
      /* Unwrap if it is a ThrowableWrapper */
      e = (e instanceof ThrowableWrapper) ? e.getCause() : e;
      /* Check if we had catch block for that type of exception */
      Object catchBlock = catches.get(e.getClass());
      if (catchBlock != null) {
        /* Bind exception */
        env.put(catchBindings.get(e.getClass()), e);
        /* Evaluate corresponding catch block */
        return evaluator.eval(catchBlock, env);
      }
      /* Unexpected exception, re-throw it */
      throw new ThrowableWrapper(e);
    } finally {
      /* And finally, evaluate finally block (if present) */
      if (fin != null) {
        evaluator.eval(fin, env);
      }
    }
  }
}

File: src/main/java/core/scm/specialforms/Try.kt
package core.scm.specialforms

import core.Evaluator
import core.Reflector
import core.Writer
import core.exceptions.IllegalSyntaxException
import core.scm.Symbol

object Try : SpecialForm("try") {

    object Catch : SpecialForm("catch") {
        override fun eval(form: List<Any?>, evaluator: Evaluator): Nothing {
            throw IllegalSyntaxException(toString(), Writer.write(form), "not allowed as an expression")
        }
    }

    object Finally : SpecialForm("finally") {
        override fun eval(form: List<Any?>, evaluator: Evaluator): Nothing {
            throw IllegalSyntaxException(toString(), Writer.write(form), "not allowed as an expression")
        }
    }

    private val reflector = Reflector()

    override fun eval(form: List<Any?>, evaluator: Evaluator): Any? {
        if (form.isEmpty()) {
            return null
        }
        var hadCatch = false
        val catches = LinkedHashMap<Class<*>, Any?>()
        var catchBindings = mutableMapOf<Class<*>, Symbol>()
        var fin: Any? = null
        val expressions = mutableListOf<Any?>()
        /* Init and check syntax */
        for (i in 1 until form.size) {
            val expr = form[i]
            if (expr is List<*> && !expr.isEmpty()) {
                val op = evaluator.env.resolve(expr.first())
                if (op == Finally) {
                    if (i != form.size - 1) {
                        throw IllegalSyntaxException("$name: finally clause must be last in try expression")
                    }
                    if (expr.size > 1) {
                        fin = listOf(Begin) + expr.drop(1)
                    }
                    continue
                } else if (op === Catch) {
                    if (expr.size < 3) {
                        throw IllegalSyntaxException("$Catch: bad syntax in form: $expr")
                    }
                    hadCatch = true
                    if (catches.isEmpty()) {
                        catchBindings = HashMap()
                    }
                    val clazz = reflector.getClazz(expr[1].toString())
                    val catchExpr = when {
                        expr.size > 3 -> listOf(Begin) + expr.drop(3)
                        else -> null
                    }
                    catches[clazz] = catchExpr
                    val sym = expr[2] as? Symbol ?:
                              throw IllegalSyntaxException("$Catch: bad binding form, expected Symbol, actual: ${expr[2]}")
                    catchBindings[clazz] = sym
                    continue
                }
            }
            if (hadCatch) {
                throw IllegalSyntaxException("$name: only catch or finally clause can follow catch in try expression")
            }
            expressions.add(expr)
        }
        /* Now Evaluate everything */
        try {
            var result: Any? = null
            expressions.forEach { result = evaluator.eval(it) }
            return result
        } catch (e: Throwable) {
            /* Check if we had catch block for that type of exception (OR for any superclass) */
            for (clazz in catches.keys) {
                if (clazz.isAssignableFrom(e.javaClass)) {
                    /* Bind exception */
                    evaluator.env[catchBindings[clazz]] = e
                    /* Evaluate corresponding catch block */
                    return evaluator.eval(catches[clazz])
                }
            }
            /* Unexpected exception, re-throw it */
            throw e
        } finally {
            /* And finally, evaluate finally block (if present) */
            fin?.let { evaluator.eval(fin) }
        }
    }
}
